### PR TITLE
perf: refactor lifecycle query

### PR DESCRIPTION
**Pull request by hazzadous**
## Changes

This change does a couple of things:

1. refactor the existing three queries as before into one. The query looks for any 
activity in the requested range, decides if it is 'new', 'returning' or 'resurrecting', and 
then checks if the period after is 'dormant'

2. Uses EventQuery query builder to build the event query, such that we can take
advantage of optimizations there.

## Further changes not implemented

_I originally had to add this in to get the events query working for 20m person_distinct_ids, but managed to reduce the pdi2 joins_

3. WACKY: uses cityHash64 to make the join onto `person_distinct_id2` smaller. For the 
lifecycle query, we're joining onto this table 4 times. When a team has a lot of distinct 
ids, this can run above the memory limit. This changes seems rather irksome however, 
if we can reduce the number of pdi2 joins we do then this would be far more desirable.
Hashing person_id doesn't produce much change. It's a UUID which I guess is already 
pretty compact. person_id is not included in these stats I don't think as they are probably 
optimized out, as are not "required keys" by the containing query.

```
      SELECT count(*)
      FROM events e
          ANY INNER JOIN (
              SELECT distinct_id,
                  argMax(person_id, version) as person_id
              FROM person_distinct_id2
              WHERE team_id = <large_team_id>
              GROUP BY distinct_id
              HAVING argMax(is_deleted, version) = 0
          ) AS pdi ON e.distinct_id = pdi.distinct_id
      WHERE team_id = <large_team_id>

Query stats:
    ──────
    event_time:        2022-01-07 11:49:36
    query_duration_ms: 25978
    read_rows:         980469550
    read_size:         69.93 GiB
    result_rows:       1
    result_size:       64.00 B
    memory_usage:      7.21 GiB

      SELECT count(*)
      FROM events e
          ANY INNER JOIN (
              SELECT cityHash64(distinct_id) as distinct_id,
                  argMax(person_id, version) as person_id
              FROM person_distinct_id2
              WHERE team_id = <large_team_id>
              GROUP BY distinct_id
              HAVING argMax(is_deleted, version) = 0
          ) AS pdi ON cityHash64(e.distinct_id) = pdi.distinct_id
      WHERE team_id = <large_team_id>

Query stats:
    ──────
    event_time:        2022-01-07 11:51:44
    query_duration_ms: 12456
    read_rows:         980481128
    read_size:         69.93 GiB
    result_rows:       1
    result_size:       64.00 B
    memory_usage:      2.45 GiB
```

4. Actually bound the unbounded events, and use created_by as the first activity. Note that there is a PR for this [here](https://github.com/PostHog/posthog/pull/7886). I've kept this separate as that change changes the functionality of the feature, and also didn't help me get this query working with my test team, but we should go ahead with this as mentioned in the referred issue

Refers to https://github.com/PostHog/posthog/issues/7382

### CODE REVIEW

File: ee/clickhouse/sql/trends/lifecycle.py [hunk 1]
```diff
@@ -1,165 +1,109 @@
...
+        dateTrunc(selected_period, selected_date_to) AS selected_interval_to,
+        selected_interval_from - INTERVAL {interval} AS previous_date_from,
+        dateDiff(
+           selected_period, 
+            -- Include the period before the first in the activity consideration
+            previous_date_from, 
+            selected_interval_to
+        ) AS number_of_intervals,
+
+        filtered_events AS (
```
Comment (macobo) [inline ee/clickhouse/sql/trends/lifecycle.py:18]: Note: Will this cause issues due with distributed tables? We've seen previously that it might.

Comment (hazzadous) [reply ee/clickhouse/sql/trends/lifecycle.py:18]: I've been testing on production ™️  and this query does work. However, yes in certain configurations of CTEs and on certain nodes we do end up not seeing some CTEs. I've added an [issue for this](https://github.com/ClickHouse/ClickHouse/issues/29748) but haven't reproduced locally, when I have time I'll try a little harder. For the time when I introduce the `EventQuery` I'll just string substitute where it's needed.

btw are you talking about the distributed table engine? I didn't think we were using that?

Comment (hazzadous) [reply ee/clickhouse/sql/trends/lifecycle.py:18]: For reference, looks like [this issue](https://github.com/ClickHouse/ClickHouse/issues/23865) might be relevant regarding CTEs, I'll update my issue

File: ee/clickhouse/sql/trends/lifecycle.py [hunk 2]
```diff
@@ -1,184 +1,129 @@
...
-        ) 
-        AND next_period != {trunc_func}(toDateTime(%(date_from)s) + INTERVAL {interval} - INTERVAL {sub_interval}))
-        OR (
-            (neighbor(person_id, -1) = person_id) 
-            AND neighbor(next_period, -1) < next_period - INTERVAL {interval}
+LIFECYCLE_SQL = f"""
+WITH 
+    %(interval)s AS selected_period,
+
+    -- enumerate all requested periods, so we can zero fill as needed.
```
Comment (macobo) [inline ee/clickhouse/sql/trends/lifecycle.py:89]: Aside: Please don't use -- in sql comments. This breaks copying queries from metabase, which hurts debugging a lot.

Comment (hazzadous) [reply ee/clickhouse/sql/trends/lifecycle.py:89]: We strip comments before exec [here](https://github.com/PostHog/posthog/blob/6fec8bcadd2b3e068833d023ba437b92ca75236a/ee/clickhouse/client.py#L241) so this should be an issue, am I missing something?

Happy to use /* */ if there is still a concern

Comment (hazzadous) [reply ee/clickhouse/sql/trends/lifecycle.py:89]: (also, re debugging and metabase, you can use export to csv or similar, which retains new lines. It's not ideal)

**Review by EDsCODE — APPROVED** [review]
Looks good. I went through the query logic and it's much clearer than previously. Did you try

**Comment by macobo** [comment]
Taking this over. Some TODOs:
- [ ] Handling created_at
- [ ] Supporting person predicate pushdowns
- [ ] Supporting materialized columns
- [ ] Removing CTEs, especially nested ones - won't work with DISTRIBUTED table engine.
- [ ] (Stretch) Adding groups support
- [ ] Make sure unbounded logic works correctly if keeping it
- [ ] Remove TrendsEventQuery inheritance - we're reading too much data for no reason

Note that the person_distinct_id optimization will not be part of this change. There's a high-level plan outlined for it in https://github.com/PostHog/product-internal/pull/240 instead.

**Comment by macobo** [comment]
Closing this for now since the PR history will be impossible to follow due to takeover and create a new PR soon! :)

Now I've dug into this, it's really well done, especially the dormant handling logic. We can speed it up further by using created_at and minimizing reading columns.

**Pull request closed without merging**